consensus/src/consensus_provider.rs

```diff
@@ -112,79 +118,87 @@ pub fn start_consensus(
+    observer_network_client: NetworkClient<ConsensusObserverMessage>,
+    observer_network_service_events: NetworkServiceEvents<ConsensusObserverMessage>,
+    consensus_publisher: Option<Arc<ConsensusPublisher>>,
     state_sync_notifier: Arc<dyn ConsensusNotificationSender>,
     consensus_to_mempool_sender: mpsc::Sender<QuorumStoreRequest>,
     aptos_db: DbReaderWriter,
     reconfig_events: Option<ReconfigNotificationListener<DbBackedOnChainConfig>>,
 ) -> Runtime {
-    let publisher_enabled = node_config.consensus_observer.publisher_enabled;
+    // Create a consensus observer runtime
     let runtime = aptos_runtimes::spawn_named_runtime("observer".into(), None);
-    let root = aptos_db.reader.get_latest_ledger_info().unwrap();
 
-    let txn_notifier = Arc::new(MempoolNotifier::new(
-        consensus_to_mempool_sender.clone(),
-        node_config.consensus.mempool_executed_txn_timeout_ms,
-    ));
+    // Create the consensus observer client
+    let consensus_observer_client = ConsensusObserverClient::new(observer_network_client.clone());
 
-    let execution_proxy = ExecutionProxy::new(
-        Arc::new(BlockExecutor::<AptosVM>::new(aptos_db)),
-        txn_notifier,
-        state_sync_notifier,
-        runtime.handle(),
-        TransactionFilter::new(node_config.execution.transaction_filter.clone()),
-    );
+    // Create the consensus observer network events
+    let observer_network_events =
+        ConsensusObserverNetworkEvents::new(observer_network_service_events);
 
+    // Create the (dummy) consensus network client
     let (self_sender, _self_receiver) =
         aptos_channels::new_unbounded(&counters::PENDING_SELF_MESSAGES);
-    let dummy_client = ConsensusNetworkClient::new(NetworkClient::new(
+    let consensus_network_client = ConsensusNetworkClient::new(NetworkClient::new(
         vec![],
         vec![],
         HashMap::new(),
         observer_network_client.get_peers_and_metadata(),
     ));
-    let bounded_executor = BoundedExecutor::new(32, runtime.handle().clone());
-    let rand_storage = Arc::new(RandDb::new(node_config.storage.dir()));
-
-    let execution_client = Arc::new(ExecutionProxyClient::new(
-        node_config.consensus.clone(),
-        Arc::new(execution_proxy),
-        AccountAddress::ONE,
-        self_sender.clone(),
-        dummy_client,
-        bounded_executor.clone(),
-        rand_storage.clone(),
-        if publisher_enabled {
-            Some(observer_network_client.clone())
-        } else {
-            None
-        },
-    ));
-
-    let events: Vec<_> = observer_network_service_events
-        .into_network_and_events()
-        .into_values()
-        .collect();
-    let network_events = Box::new(select_all(events));
-
-    let reconfig_events =
-        reconfig_events.expect("Reconfig events are required for the consensus observer!");
 
+    // If the consensus observer is enabled, create the execution client.
+    // If not, stub it out with a dummy client.
```

Comment:
why do we still create observer when it's not enabled? do we want to support observer -> non-observer -> publisher path?

Reply:
> why do we still create observer when it's not enabled?

Aah, this is because the current implementation is a little messy. I'll clean it up on the next pass. But:
- If the publisher is enabled (and not the observer), we need the publisher to take the network events handle so that it can receive (and respond to) subscription event messages.
- If the observer is enabled (and not the publisher), it needs the network events handle to receive the published messages.
- If both are enabled, then the network handle needs to be used by both (which currently isn't really possible, e.g., no clone support to pass to both components, etc.). To avoid this (for now(, I've made it so that the observer always takes the handle and then passes any relevant messages to the publisher, e.g., direct sends are consumed by the observer, and RPCs are sent to the publisher. But, this means that if a node doesn't enable the publisher, it still creates a dummy observer to pass it messages. I'm going to avoid this on the next pass, e.g., maybe split the handles on message types, or create a third component for message passing 😄